src/snowflake/snowpark/_internal/analyzer/datatype_mapper.py

```diff
@@ -213,9 +213,27 @@ def schema_expression(data_type: DataType, is_nullable: bool) -> str:
         else:
             return "to_timestamp('2020-09-16 06:30:00')"
     if isinstance(data_type, ArrayType):
+        if data_type.structured:
+            element = schema_expression(data_type.element_type, is_nullable)
+            return f"to_array({element}) :: {convert_sp_to_sf_type(data_type)}"
         return "to_array(0)"
     if isinstance(data_type, MapType):
+        if data_type.structured:
+            key = schema_expression(data_type.key_type, is_nullable)
+            value = schema_expression(data_type.value_type, is_nullable)
+            return f"object_construct_keep_null({key}, {value}) :: {convert_sp_to_sf_type(data_type)}"
```

Comment:
should we determine whether keeping null values based on `is_nullable`?

Reply:
If we don't keep nulls and either the key or value gets evaluated to a `NULL :: ` type statement then that field would be dropped from the schema altogether. For this reason I think we always want nulls.